src/librbd/cache/rwl/Request.cc

```diff
@@ -0,0 +1,361 @@
+  m_resources.buffers.reserve(image_extents_size);
+
+  bytes_cached = 0;
+  bytes_allocated = 0;
+  number_lanes = image_extents_size;
+  number_log_entries = image_extents_size;
+  number_unpublished_reserves = image_extents_size;
+
+  for (auto &extent : this->image_extents) {
+    m_resources.buffers.emplace_back();
+    struct WriteBufferAllocation &buffer = m_resources.buffers.back();
+    buffer.allocation_size = MIN_WRITE_ALLOC_SIZE;
+    buffer.allocated = false;
+    bytes_cached += extent.second;
+    if (extent.second > buffer.allocation_size) {
+      buffer.allocation_size = extent.second;
+    }
+    bytes_allocated += buffer.allocation_size;
+  }
+  bytes_dirtied = bytes_cached;
+}
+
+template <typename T>
+void C_WriteRequest<T>::setup_log_operations() {
+  {
+    std::lock_guard locker(m_lock);
+    // TODO: Add sync point if necessary
+    std::shared_ptr<SyncPoint> current_sync_point = rwl.get_current_sync_point();
+    uint64_t current_sync_gen = rwl.get_current_sync_gen();
+    op_set =
+      make_unique<WriteLogOperationSet>(this->m_dispatched_time,
+                                        m_perfcounter,
+                                        current_sync_point,
+                                        rwl.get_persist_on_flush(),
+                                        rwl.get_context(), this);
+    ldout(rwl.get_context(), 20) << "write_req=" << *this << " op_set=" << op_set.get() << dendl;
+    ceph_assert(m_resources.allocated);
+    /* op_set->operations initialized differently for plain write or write same */
+    for (auto &extent : this->image_extents) {
+      /* operation->on_write_persist connected to m_prior_log_entries_persisted Gather */
+      auto operation =
+        std::make_shared<WriteLogOperation>(*op_set, extent.first, extent.second, rwl.get_context());
+      op_set->operations.emplace_back(operation);
+    }
+    auto allocation = m_resources.buffers.begin();
+    uint64_t buffer_offset = 0;
+    for (auto &gen_op : op_set->operations) {
+      /* A WS is also a write */
+      auto operation = gen_op->get_write_op();
+      ldout(rwl.get_context(), 20) << "write_req=" << *this << " op_set=" << op_set.get()
+                                   << " operation=" << operation << dendl;
+      rwl.inc_last_op_sequence_num();
+      operation->init(true, allocation, current_sync_gen,
+                      rwl.get_last_op_sequence_num(), this->bl, buffer_offset, op_set->persist_on_flush);
+      buffer_offset += operation->log_entry->write_bytes();
+      ldout(rwl.get_context(), 20) << "operation=[" << *operation << "]" << dendl;
+      allocation++;
+    }
+  }
+    /* All extent ops subs created */
+  op_set->extent_ops_appending->activate();
+  op_set->extent_ops_persist->activate();
+
+  /* Write data */
+  for (auto &operation : op_set->operations) {
+    /* operation is a shared_ptr, so write_op is only good as long as operation is in scope */
+    auto write_op = operation->get_write_op();
```

Comment:
Nit: drop `get_write_op` and replace w/ a generic handler method that can be properly overridden in the write op derived class.

Reply:
Done